README.md
# record-a-cast [![Build status](https://travis-ci.org/twolfson/record-a-cast.svg?branch=master)](https://travis-ci.org/twolfson/record-a-cast)

Select and record a portion of your desktop

`record-a-cast` was created out of frustration for either poor UI, poor compatibility with Linux, or both. It's based on well-tested technologies (e.g. [Electron][], [FFmpeg][]) so it should work well on most environments.

[Electron]: https://github.com/atom/electron
[FFmpeg]: https://www.ffmpeg.org/

## Demonstration
**1) Select an area**

![Select an area image][]

[Select an area image]: http://i.imgur.com/hOIceGa.png

**2) Get a screencast**

![Get a screencast image][]

[Get a screencast image]: http://i.imgur.com/jO8vvMa.gif

## Donating
Support this project and [others by twolfson][twolfson-projects] via [PayPal][paypal-twolfson].

[![Support via PayPal][paypal-button]][paypal-twolfson]

[twolfson-projects]: http://twolfson.com/projects
[paypal-button]: http://rawgit.com/twolfson/paypal-github-button/1.x.x/dist/button.svg
[paypal-twolfson]: http://bit.ly/twolfson-paypal-5

## Requirements
- X11 server, [FFmpeg][] requires this for its `x11grab` functionality
- [npm][], usually installed with [Node.js][]
    - Used for installing dependencies
    - We recommend installing via: <https://github.com/nodesource/distributions>
- [FFmpeg][], used for recording the screen
    - This might be known as `avconv` on your distribution
    - For `apt` installation, this might be one of:
        - `apt-get install ffmpeg`
        - `apt-get install libav-tools`

[npm]: http://npmjs.org/
[Node.js]: http://nodejs.org/

## Getting Started
1) Install the module globally via `npm install -g record-a-cast`

2) Start running `record-a-cast` via `record-a-cast recording.mov`

3) Select an area via a draggable/resizable window

![Select an area image][]

4) Press `Enter` and FFmpeg will begin recording that area

5) To stop recording, send a keyboard interrupt to `record-a-cast` (typically `Ctrl+C`)

6) Your recording will be available at `recording.mov`

![Recording image][Get a screencast image]

## Documentation
### CLI usage
Our CLI currently supports the following:

```
$ record-a-cast --help

  Usage: record-a-cast [options] <outfile> [-- ffmpeg-options]

  Options:

    -h, --help                  output usage information
    -V, --version               output the version number
    --delay <delay>             Milliseconds to wait before recording starts
                       (default: 50 to prevent visible overlays)
    --height-divisor <divisor>  Resize only to heights divisible by `m`
    --width-divisor <divisor>   Resize only to widths divisible by `n`

```

As noted, we support passing through options to FFmpeg via a `--` delimiter. For example, we can pass `-r 10` (record at 10FPS) to `ffmpeg`:

```bash
record-a-cast out.mov -- -r 10
# Invokes: /usr/bin/ffmpeg -video_size 20x20 -f x11grab -i :0+20,20 -y -r 10 out.mov
```

### Environment variables
We support the following environment variables:

- FFMPEG_BIN `String` - Path to desired FFmpeg executable
    - By default, we will search for `ffmpeg` and `avconv` on `PATH`

## Examples
### Custom height/width divisors
On some computers, we can encounter errors like:

```
[libx264 @ 0x12f24c0] width not divisible by 2 (801x600)
```

These are caused by our codec requiring even heights/widths for its format. To work around this, we support snapping our selection to height/widths that are only even (e.g. `801 -> 802`).

To run `record-a-cast` with snapping to even heights/widths, use:

```bash
record-a-cast --height-divisor 2 --width-divisor 2--out.mov
```

### Custom framerate
When recording something like a GIF, we might want a reduced framerate to save on size and poor frame delays. Here's an example of recording a movie at 10FPS:

```bash
record-a-cast out.mov -- -r 10
# Invokes: /usr/bin/ffmpeg -video_size 20x20 -f x11grab -i :0+20,20 -y -r 10 out.mov
```

### Duration
Sometimes we might want to only record for a few seconds. In this example, we will set our framerate to 24FPS and record for 3 seconds (72 frames).

```bash
record-a-cast out.mov -- -r 24 -frames 72
# Invokes: /usr/bin/ffmpeg -video_size 20x20 -f x11grab -i :0+20,20 -y -r 24 -frames 72 out.mov

# Compute duration dynamically in bash
# record-a-cast out.mov -- -r 24 -frames "$((3 * 24))"
```

### Creating a GIF
**Requirements:**

- [ImageMagick][], image manipulation library which is much better at generating GIFs than FFmpeg
    - `apt-get install imagemagick`

[ImageMagick]: http://www.imagemagick.org/script/index.php

**Script:**

```bash
# Record our screencast with a reduced framerate
# DEV: We use `-r 10` to output 10FPS since GIFs move slower
record-a-cast recording.mov -- -r 10

# Output each frame to a folder
# DEV: When we tried to record directly to PNGs via FFmpeg, they were all empty
#   Hence, this 2 step process
mkdir frames
ffmpeg -i recording.mov frames/recording%03d.png

# DEV: If you want to remove any frames from the final product
#   then navigate to `frames` and delete them

# Combine our frames into a GIF via ImageMagick
convert -loop 0 frames/recording*.png recording.gif
```

Want more variations on this? See the following links:

- https://gist.github.com/dergachev/4627207
- https://gist.github.com/SlexAxton/4989674

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint via `npm run lint` and test via `npm test`.

## Unlicense
As of Mar 17 2016, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the [UNLICENSE][].

[UNLICENSE]: UNLICENSE


CHANGELOG.md
# record-a-cast changelog
1.5.2 - Added back missed shell-quote@1.7.3 upgrade

1.5.1 - Removed Node.js@18 from Travis CI

1.5.0 - Upgraded to shell-quote@1.7.3 to fix GitHub vulnerability warning and bumped to Node.js>=14

1.4.2 - Upgraded to shell-quote@1.6.1 to fix GitHub vulnerability warning

1.4.1 - Moved "Donating" section towards top in README

1.4.0 - Added support for height/width divisors

1.3.0 - Added support for `--delay`

1.2.0 - Added argument passing for FFmpeg

1.1.0 - Added arrow key support

1.0.1 - Added notes for manual testing

1.0.0 - Initial release


test/record-a-gif-test.js
// Load in dependencies
var assert = require('assert');
var recordAGif = require('../');

// Start our tests
describe('record-a-gif', function () {
  it('returns awesome', function () {
    assert.strictEqual(recordAGif(), 'awesome');
  });
});


lib/get-geometry.js
// Load in our dependencies
var BrowserWindow = require('electron').BrowserWindow;

// Define our function
function getGeometry(cb) {
  // Create a new transparent window
  // TODO: Occupy full viewport?
  var browserWindow = new BrowserWindow({transparent: true, frame: false});
  browserWindow.loadURL('file://' + __dirname + '/browser/get-geometry.html');

  // Create a funciton to update our
  var bounds = browserWindow.getBounds();
  console.log(bounds);

  // When our window is resized, update its bounds
  browserWindow.on('resize', function handleResize () {
    bounds = browserWindow.getBounds();
    console.log(bounds);
  });

  // When the window closes, callback with our bounds
  browserWindow.on('close', function handleClose () {
    cb(null, bounds);
  });
}

// Export our function
module.exports = getGeometry;


lib/record-a-gif.js
// Load in our dependencies
var app = require('electron').app;
var program = require('commander');
var getGeometry = require('./get-geometry');

// Load in package info
var pkg = require('../package.json');

// Handle CLI arguments
program
  .version(pkg.version)
  // Allow unknown Chromium flags
  // https://github.com/atom/electron/blob/v0.26.0/docs/api/chrome-command-line-switches.md
  .allowUnknownOption();
program.parse(process.argv);

// When Electron is done loading, launch our application
app.on('ready', function handleReady () {
  // TODO: Allow receiving of geometry from CLI
  // getGeometry(function handleGeometry (err, geometry) {
  //   // If there was an error, then throw it
  //   if (err) {
  //     throw err;
  //   }
  // TODO: Remove dev next tick
  process.nextTick(function handleNextTick () {
    var geometry = {x: 1520, y: 227, width: 800, height: 600};

    // Otherwise, log the geometry
    console.log('got geometry', geometry);

    // Start a process to capture a video
    // TODO: Record and compare the following
    //  avconv recorded GIF
    //    avconv -video_size 300x200 -f x11grab -i "$DISPLAY"+100,200 -r 10 -frames 30 -pix_fmt rgb24 gif.gif
    //  avconv recorded MP4 that's converted to GIF via `avconv`
    //    avconv -video_size 300x200 -f x11grab -i "$DISPLAY"+100,200 -r 10 -frames 30 mp4.mp4
    //    avconv -i mp4.mp4 -vf palettegen palette.png
    //      https://ffmpeg.org/ffmpeg-filters.html#palettegen-1
    //    libav-tools seems to lack `palettegen` so this is not plausible
    //  avconv recorded MP4 that's converted to GIF via ImageMagick
    //    avconv -video_size 300x200 -f x11grab -i "$DISPLAY"+100,200 -r 10 -frames 30 mp4.mp4
    //    avconv -i mp4.mp4 frames/out%03d.png
    //    convert -loop 0 frames/*.png imagemagick.gif

    // TODO: For easier development, set `-frames` count and allow this to run on loop
    //   prob name the CLI parameter `--duration`

    // TODO: Set low frame rate like 10
    // Attribution to: https://gist.github.com/dergachev/4627207
    //   and https://trac.ffmpeg.org/wiki/Capture/Desktop
    // Never mind this is a literal GIF output and the colors are effed
    //   avconv -video_size 1024x768 -f x11grab -i "$DISPLAY"+100,200 -pix_fmt rgb24 -f gif - | cat
    // This is generating transparent images for some reason...
    //   avconv -video_size 1024x768 -f x11grab -i "$DISPLAY"+100,200 frames/ffout%03d.png
    // This works but mp4... =_=
    //   avconv -video_size 1024x768 -f x11grab -i "$DISPLAY"+100,200 out.mp4
  });
});


lib/select-geometry.js
// Load in our dependencies
var assert = require('assert');
var app = require('electron').app;
var BrowserWindow = require('electron').BrowserWindow;
var ipcMain = require('electron').ipcMain;

// If we are on Linux, then add transparency flags
// https://github.com/atom/electron/blob/v0.37.2/docs/api/frameless-window.md#limitations
// https://github.com/atom/electron/blob/v0.37.2/docs/api/app.md#appcommandlineappendswitchswitch-value
if (process.platform === 'linux') {
  app.commandLine.appendSwitch('enable-transparent-visuals');
  app.commandLine.appendSwitch('disable-gpu');
}

// Define a rounding helper (e.g. `(1, 2) -> 2`, `(0, 2) -> 2`, `(3, 2) -> 4`, `(-3, 2) -> -4`)
function round(val, divisor) {
  // Calculate the remainder (e.g. `3 % 2 === 1`)
  var negativeDivisor = val < 0 ? -1 : 1;
  val *= negativeDivisor;
  var remainder = val % divisor;

  // If the remainder is closer to the divisor, then add the remaining amount to our value
  // DEV: We use multiplication to avoid floating point issues
  if (remainder * 2 >= divisor) {
    return negativeDivisor * (val - remainder + divisor);
  // Otherwise, return the value without our remainder
  } else {
    return negativeDivisor * (val - remainder);
  }
}

// Define our function
function selectGeometry(mainParams, cb) {
  // Verify we received width divisor and height divisor
  var heightDivisor = mainParams.heightDivisor;
  var widthDivisor = mainParams.widthDivisor;
  assert(widthDivisor && !isNaN(widthDivisor));
  assert(heightDivisor && !isNaN(heightDivisor));

  // Create a new transparent window
  // TODO: Occupy full viewport like Shutter?
  var browserWindow = new BrowserWindow({
    frame: false,
    height: round(600, heightDivisor),
    width: round(800, widthDivisor),
    transparent: true
  });
  browserWindow.loadURL('file://' + __dirname + '/browser/select-geometry.html');

  // Save our window's current bounds
  // https://github.com/atom/electron/blob/v0.37.2/docs/api/browser-window.md#wingetbounds
  var bounds;
  function updateBounds() {
    bounds = browserWindow.getBounds();
  }
  updateBounds();

  // When our window is resized, update its bounds
  browserWindow.on('resize', function handleResize() {
    // Get the resized bounds
    // bounds = {x, y, width, height} = {x: 20, y: 30, width: 400, height: 500}
    var bounds = browserWindow.getBounds();

    // Snap bounds to closest divisor and save them
    bounds.width = round(bounds.width, widthDivisor);
    bounds.height = round(bounds.height, heightDivisor);
    browserWindow.setBounds(bounds);

    // Update our bounds
    // DEV: We don't directly write as we could go outside of the desktop bounds
    //   but Electron autocorrects this :tada:
    updateBounds();
  });

  // When a window movement is requested
  function handleMoveBy(evt, params) {
    // Move our window by the requested amount
    // https://github.com/atom/electron/blob/v0.37.2/docs/api/browser-window.md#winsetpositionx-y-animate
    // params = {x: +1, y: -2}
    // currentPosition = [x, y] = [100, 200]
    var currentPosition = browserWindow.getPosition();
    browserWindow.setPosition(currentPosition[0] + params.x, currentPosition[1] + params.y);

    // Update our bounds
    updateBounds();
  }
  ipcMain.on('moveBy', handleMoveBy);

  // When a window resizing is requested
  function handleResizeBy(evt, params) {
    // Resize our window by the requested amount
    // https://github.com/atom/electron/blob/v0.37.2/docs/api/browser-window.md#winsetpositionx-y-animate
    // params = {top: +1, left: +2, right: +2, bottom: +3}
    // bounds = {x, y, width, height} = {x: 20, y: 30, width: 400, height: 500}
    var bounds = browserWindow.getBounds();
    /* eslint-disable max-statements-per-line */
    // DEV: Our directions are inverse to CSS directions so + means increase in that direction, not decrease
    var amt;
    if (params.top) { amt = round(params.top, heightDivisor); bounds.y -= amt; bounds.height += amt; }
    if (params.bottom) { bounds.height += round(params.bottom, heightDivisor); }
    if (params.left) { amt = round(params.left, widthDivisor); bounds.x -= amt; bounds.width += amt; }
    if (params.right) { bounds.width += round(params.right, widthDivisor); }
    browserWindow.setBounds(bounds);
    /* eslint-enable max-statements-per-line */

    // Update our bounds
    // DEV: We don't directly write as we could go outside of the desktop bounds
    //   but Electron autocorrects this :tada:
    updateBounds();
  }
  ipcMain.on('resizeBy', handleResizeBy);

  // Set up drag handlers (using click handlers in the window)
  var dragStartPointerPosition; // eslint-disable-line one-var
  function calculateAbsolutePosition(browserPosition, pointerPosition) {
    /*
    +---------------+
    |               |
    |   +-------+   |
    |   |  *    |   |
    |   |       |   |
    |   +-------+   |
    |               |
    +---------------+
    */
    // browserPosition = [x, y] = [100, 200]
    // pointerPosition = {x, y} = {x: 20, y: 10}
    // absolutePosition = {x, y} = {x: 120, y: 210}
    return {
      x: browserPosition[0] + pointerPosition.x,
      y: browserPosition[1] + pointerPosition.y
    };
  }
  // When a drag starts, save the current pointer offset
  function handleDragStart(evt, data) {
    dragStartPointerPosition = data;
  }
  ipcMain.on('dragStart', handleDragStart);
  // When a drag moves
  function handleDragMove(evt, data) {
    // Verify we have a start position
    assert(dragStartPointerPosition);

    // Determine the current absolute position of the cursor
    // https://github.com/atom/electron/blob/v0.37.2/docs/api/browser-window.md#winsetpositionx-y-animate
    var currentAbsolutePosition = calculateAbsolutePosition(browserWindow.getPosition(), data);

    // Calculate the new position of our window so the cursor stays in place
    //   and goes back to its original offset
    // DEV: dragStartPointerPosition is the position of the cursor with respect to the window's vieport
    //   when we started dragging
    var newX = currentAbsolutePosition.x - dragStartPointerPosition.x;
    var newY = currentAbsolutePosition.y - dragStartPointerPosition.y;

    // Update our location and save our bounds
    browserWindow.setPosition(newX, newY);
    updateBounds();
  }
  ipcMain.on('dragMove', handleDragMove);
  // When a drag ends, clean up
  function handleDragEnd(evt, data) {
    dragStartPointerPosition = null;
  }
  ipcMain.on('dragEnd', handleDragEnd);

  // When the window closes
  browserWindow.on('closed', function handleClosed() {
    // Remove our IPC subscriptions
    ipcMain.removeListener('dragStart', handleDragStart);
    ipcMain.removeListener('dragMove', handleDragMove);
    ipcMain.removeListener('dragEnd', handleDragEnd);
    ipcMain.removeListener('moveBy', handleMoveBy);
    ipcMain.removeListener('resizeBy', handleResizeBy);

    // Callback with our bounds
    cb(null, bounds);
  });
}

// Export our function
module.exports = selectGeometry;


lib/record-a-cast.js
// When an unknown exception occurs, fully bail
// DEV: By default, Electron will log/alert but keep on running
process.on('uncaughtException', function handleUncaughtException(err) {
  throw err;
});

// Load in our dependencies
var app = require('electron').app;
var async = require('async');
var program = require('commander');
var shellQuote = require('shell-quote').quote;
var which = require('which');
var spawn = require('child_process').spawn;
var selectGeometry = require('./select-geometry');

// Resolve our FFmpeg path
var FFMPEG_BIN = process.env.FFMPEG_BIN;
if (FFMPEG_BIN === undefined) {
  /* eslint-disable no-empty */
  try { FFMPEG_BIN = which.sync('ffmpeg'); } catch (err) {}
  if (FFMPEG_BIN === undefined) {
    try { FFMPEG_BIN = which.sync('avconv'); } catch (err) {}
  }
  /* eslint-enable no-empty */
}
if (FFMPEG_BIN === undefined) {
  throw new Error('Failed to find `ffmpeg` or `avconv`. ' +
    'Please verify either one is installed and on the `PATH` environment variable ' +
    'or use the `FFMPEG_BIN` environment variable');
}

// Resolve our current DISPLAY
var DISPLAY = process.env.DISPLAY;
if (DISPLAY === undefined) {
  throw new Error('Expected `DISPLAY` environment variable to be defined but it was not. ' +
    'This is required for FFmpeg\'s `x11grab`, please define it');
}

// Load in package info
var pkg = require('../package.json');

// When all windows close, keep our app open
app.on('window-all-closed', function handleWindowAllClosed() {
  // Do nothing, this listener keeps our app open
});

// Define helper for parsing CLI
function parseInt10(str) {
  var val = parseInt(str, 10);
  if (isNaN(val)) {
    throw new Error('"' + str + '" could not be parsed as an integer');
  }
  return val;
}

// Set up CLI parser
program.name = pkg.name;
program.version(pkg.version);

// Set up our main command
var mainCalled = false;
program
  .usage('[options] <outfile> [-- ffmpeg-options]')
  .option('--delay <delay>', 'Milliseconds to wait before recording starts\n' +
            '                   (default: 50 to prevent visible overlays)', parseInt10, 50)
  .option('--height-divisor <divisor>', 'Resize only to heights divisible by `m`', parseInt10, 1)
  .option('--width-divisor <divisor>', 'Resize only to widths divisible by `n`', parseInt10, 1);

program.action(function main(outfile/* , userFfmpegArgs..., program */) {
  // Set `mainCalled` to true
  mainCalled = true;

  // Extract our user's ffmpegArgs
  var userFfmpegArgs = [].slice.call(arguments, 1, -1);

  // When Electron is done loading, launch our application
  app.on('ready', function handleReady() {
    // Set up a SIGINT handler to fully quit
    // DEV: This is to prevent any FFmpeg starting while we kill our process
    var receivedSIGINT = false;
    function handleSIGINT() {
      console.error('SIGINT received. Exiting...');
      receivedSIGINT = true;
      process.exit(1);
    }
    process.on('SIGINT', handleSIGINT);

    // In series
    var geometry;
    async.waterfall([
      function selectGeometryFn(cb) {
        selectGeometry({
          heightDivisor: program.heightDivisor,
          widthDivisor: program.widthDivisor
        }, cb);
      },
      function handleSelectGeometry(_geometry, cb) {
        // If we received a SIGINT, then stop
        if (receivedSIGINT) {
          return;
        }

        // Log and save our requested geometry
        console.log('Requested geometry:', _geometry);
        geometry = _geometry;
        cb();
      },
      function waitForDelay(cb) {
        /* eslint-disable callback-return */
        // If there is a delay, then wait for it
        if (program.delay) {
          console.log('Waiting for ' + program.delay + 'ms before recording...');
          setTimeout(function handleSetTimeout() {
            cb();
          }, program.delay);
        // Otherwise, wait for the next tick (to prevent Zalgo)
        } else {
          process.nextTick(cb);
        }
        /* eslint-enable callback-return */
      },
      function recordFfmpeg() {
        // Remove our SIGINT handler
        process.removeListener('SIGINT', handleSIGINT);

        // If we received a SIGINT, then stop
        if (receivedSIGINT) {
          return;
        }

        // Set up our process invocation
        // https://trac.ffmpeg.org/wiki/Capture/Desktop
        var ffmpegCmd = FFMPEG_BIN;
        var ffmpegArgs = [
          // Specify record area dimensions
          //   -video_size 300x200
          '-video_size', geometry.width + 'x' + geometry.height,
          // Indicate our source
          '-f', 'x11grab',
          //   :0.0+100,200
          '-i', DISPLAY + '+' + geometry.x + ',' + geometry.y,
          // Overwrite existing files
          // DEV: If we want to support overwrite prompting, then do it in Electron
          '-y'
          // Concatenate user input arguments
        ].concat(userFfmpegArgs).concat([
          // Indicate our output
          outfile
        ]);

        // Log our information
        console.log('Starting FFmpeg via:', shellQuote([ffmpegCmd].concat(ffmpegArgs)));

        // Start our child process
        // DEV: SIGINT will automatically propagate to the child
        var child = spawn(ffmpegCmd, ffmpegArgs, {stdio: 'inherit'});

        // When the child exits, we exit with the same code
        child.on('close', function handleClose(code, signal) {
          process.exit(code);
        });
      }
    ], function handleError(err) {
      // If there was an error, then throw it
      if (err) {
        throw err;
      }
    });
  });
});

// Parse our CLI
program.parse(process.argv);

// If main wasn't called, then display our help and exit
if (!mainCalled) {
  program.outputHelp();
  process.exit(0);
}
